File: lightsabers/case.rb
# General Grevous
class Kaleesh < ActiveRecord::Base
  has_many :lightsabers, as: :warrior
end

class Lightsaber < ActiveRecord::Base
  belongs_to :warrior, polymorphic: true, after_add: :set_color

  def set_color(warrior)
    case warrior.class
    when Jedi
      self.color = ['green', 'blue', 'purple'].sample
    when Sith
      self.color = 'red'
    when Kaleesh
      self.color = ['green', 'blue'].sample
    end
  end
end

File: csharp/lightsabers/case.cs
"refactor.cs"

// General Grevous
class Kaleesh {
}

class Lightsaber
{
  string[] JediColors = { "green", "blue", "purple" };
  string[] KaleeshColors = { "green", "blue" };
  string SithColors = "red";

  public void AfterAdd (object klass)
  {
    switch(klass.getType())
    {
      case Jedi:
        this.color = JediColors[random.Next(0, JediColors.length)];
        break;
      case Sith:
        this.color = SithColors;
        break;
      case Kaleesh:
        this.color = KaleeshColors[random.Next(0, KaleeshColors.length)];
        break;
    }
  }
}
